### PR TITLE
Http data structures improvements, part 1

### PR DESCRIPTION
**Pull request by DanSimon**
_Notice this is to merge into a new `develop-0.7.1` branch, so this is not going into 0.7.0_

Made a bunch of improvements that both clean things up a lot and considerably improve performance.

1.  Previously headers were handled differently in requests versus responses, mainly because things were optimized for parsing requests and encoding responses.  Now there's a single `HttpHeader` trait that abstracts all that away.  Under the hood there's still two versions, one optimized for decoding and the other encoding, but now it can all be hidden from the user.
2.  Replacing the `Vector` with a plain old `Array` for the collection of headers resulted in a huge performance boost.  I tried a few different approaches to try to keep using a `Seq` or `IndexedSeq` but this one worked best.  So again to hide this detail from users and keep the API as an immutable collection, there's a new `HttpHeaders` class, used in both requests and responses.
3.  Basically did the same thing for the Http Body as well.

This isn't a complete unification but it does go a long way and this branch was starting to get pretty big.  So I'll save the rest for another PR.

### CODE REVIEW

File: colossus-tests/src/test/scala/colossus/protocols/http/client/HttpResponseParserSpec.scala [hunk 1]
```diff
@@ -35,12 +35,18 @@ class HttpResponseParserSpec extends WordSpec with MustMatchers {
...
+      val expected = HttpResponse(
         HttpVersion.`1.1`,
         HttpCodes.OK,
-        Vector("host"->"api.foo.bar:444", "accept"->"*/*", "authorization"->"Basic XXX", "accept-encoding"->"gzip, deflate", "content-length"->"0")
-      ))
-      parser.parse(DataBuffer(ByteString(res))) must equal(Some(expected))
+        Seq("host"->"api.foo.bar:444", "accept"->"*/*", "authorization"->"Basic XXX", "accept-encoding"->"gzip, deflate", "content-length"->"0")
+      )
+      val actual = parser.parse(DataBuffer(ByteString(res))).get.value
+      println(actual.head.headers.toSeq.toSet.toString)
```
Comment (nsauro) [inline colossus-tests/src/test/scala/colossus/protocols/http/client/HttpResponseParserSpec.scala]: probably want to remove these printlns.

File: colossus/src/main/scala/colossus/protocols/http/Header.scala [hunk 2]
```diff
@@ -62,37 +64,96 @@ object HttpHeaders {
...
-  def headers : Seq[(String, String)]
+  def toEncodedHeader = new EncodedHeader(encoded, key.length, key.length + 2)
+}
+
+
+object HttpHeader {
+  def apply(key: String, value: String): HttpHeader = (new DecodedHeader(key, value)).toEncodedHeader
 
 
+  object Conversions {
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/protocols/http/Header.scala]: I like this idea.  Does it make sense to look into one that is a conversion for (String, Any), which just invokes .toString on Any?

File: colossus/src/main/scala/colossus/protocols/http/Header.scala [hunk 3]
```diff
@@ -62,37 +64,96 @@ object HttpHeaders {
...
+    implicit def stringTuple2Header(t: (String, String)): HttpHeader = HttpHeader(t._1, t._2)
+    implicit def seqStringTuple2Headers(t: Seq[(String, String)]): HttpHeaders = new HttpHeaders(t.map{stringTuple2Header}.toArray)
+  }
+
+}
+    
+class HttpHeaders(private val headers: Array[HttpHeader]) {
   def singleHeader(name: String): Option[String] = {
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/protocols/http/Header.scala]: I think i prefer the name 'firstHeader', it's a bit more specific to what this function is doing.

File: colossus/src/main/scala/colossus/protocols/http/Header.scala [hunk 4]
```diff
@@ -62,37 +64,96 @@ object HttpHeaders {
...
     val l = name.toLowerCase
-    headers.collectFirst{ case (`l`, v) => v}
+    headers.collectFirst{ case x if (x.key == l) => x.value }
   }
 
   def multiHeader(name: String): Seq[String] = {
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/protocols/http/Header.scala]: More opinionated, but i think the name 'headerValues' is a bit more descriptive.

File: colossus/src/main/scala/colossus/protocols/http/HttpRequest.scala [hunk 5]
```diff
@@ -3,13 +3,77 @@ package protocols.http
 
 import akka.util.ByteString
 
-case class HttpRequest(head: HttpHead, entity: Option[ByteString]) {
+case class HttpRequestHead(method: HttpMethod, url: String, version: HttpVersion, headers: HttpHeaders) {
+
+  lazy val (path, query) = {
```
Comment (nsauro) [inline colossus/src/main/scala/colossus/protocols/http/HttpRequest.scala]: Can we put a type on this, or at least make it a case class

File: colossus/src/main/scala/colossus/protocols/http/Header.scala (hunk 2 above)
Comment (DanSimon) [inline colossus/src/main/scala/colossus/protocols/http/Header.scala]: This is just old code that got moved around.  I think allowing `Any` leaves a lot of room for people to make mistakes, but I do think there might be room for something more flexible than just string.

File: colossus/src/main/scala/colossus/protocols/http/Header.scala (hunk 4 above)
Comment (DanSimon) [inline colossus/src/main/scala/colossus/protocols/http/Header.scala]: I actually don't think we need to repeat "header" in the name, but maybe the names `firstValue` and `allValues` might be better.

File: colossus/src/main/scala/colossus/protocols/http/HttpRequest.scala (hunk 5 above)
Comment (DanSimon) [inline colossus/src/main/scala/colossus/protocols/http/HttpRequest.scala]: It's just extracting the url into two lazy vals.  Again old moved code.

**Pull request merged**